src/metabase/mbql/normalize.cljc

```diff
@@ -639,25 +639,40 @@
 
 (defn- canonicalize-mbql-clauses
   "Walk an `mbql-query` an canonicalize non-top-level clauses like `:fk->`."
-  [mbql-query]
-  (walk/prewalk
-   (fn [x]
-     (cond
-       (map? x)
-       (m/map-vals canonicalize-mbql-clauses x)
-
-       (not (mbql-clause? x))
-       x
-
-       :else
-       (try
-         (canonicalize-mbql-clause x)
-         (catch #?(:clj Throwable :cljs js/Error) e
-           (log/error (i18n/tru "Invalid clause:") x)
-           (throw (ex-info (i18n/tru "Invalid MBQL clause: {0}" (ex-message e))
-                           {:clause x}
-                           e))))))
-   mbql-query))
+  [form]
+  (cond
+    (record? form)
+    (reduce-kv (fn [r k x] (assoc r k (canonicalize-mbql-clauses x))) form form)
+
+    (map? form)
+    (m/map-vals canonicalize-mbql-clauses form)
+
+    (mbql-clause? form)
+    (let [top-canonical
+          (try
+            (canonicalize-mbql-clause form)
+            (catch #?(:clj Throwable :cljs js/Error) e
+              (log/error (i18n/tru "Invalid clause:") form)
+              (throw (ex-info (i18n/tru "Invalid MBQL clause: {0}" (ex-message e))
+                              {:clause form}
+                              e))))]
+      (if (seq top-canonical)
+        (into (conj (empty top-canonical) (first top-canonical))
+              (map canonicalize-mbql-clauses)
+              (rest top-canonical))
+        top-canonical))
+
+    (list? form)
+    (apply list (map canonicalize-mbql-clauses form))
+
+    (seq? form)
+    (doall (map canonicalize-mbql-clauses form))
```

Comment:
I think for both lists and `seq?`s  it's probably fine to do `mapv`. With `doall` or `apply list`, you're realizing everything in memory at once anyway, might as well do things the easy way if that's the case

Reply:
I did this purely for compatibility with the previous version that wouldn't change the types of these collections, but I gladly simplify these things since we are making things canonical, so changes are to be expected anyway.